cmd/dfdaemon/app/root.go

```diff
@@ -41,45 +43,107 @@ var rootCmd = &cobra.Command{
 	Long:              "The dfdaemon is a proxy between container engine and registry used for pulling images.",
 	DisableAutoGenTag: true, // disable displaying auto generation tag in cli docs
 	RunE: func(cmd *cobra.Command, args []string) error {
-		initOption(opt)
+		if viper.GetBool("version") {
```

Comment:
I'm all for clean code, but I don't think putting these code into one function necessarily makes it clearer. (I made some more changes, and added some tests for `cmd/dfdaemon/app`. It should be better than the last time you reviewed.)

The original `initOption(opt)` is not good mainly because it has unclear side effects and does much more than its name suggests. It didn't make the code clearer or easier to test.

My current implementation has 3 crucial part:
1. `getConfigFromViper`
2. `initDfdaemon`: set up logging, local repo, GOMAXPROC etc. (maybe we can choose a better function name)
3. `dfdaemon.NewFromConfig` and `Start`

The parts themselves are self-contained and testable.

```go
viper.SetConfigFile(viper.GetString("config"))
viper.SetConfigType("yaml")
if err := viper.ReadInConfig(); err != nil {
	return errors.Wrap(err, "read config file")
}
```

These config file related actions are only used in rootCmd once. Wrapping them into one function saves 2 lines here but is hardly any clearer.

The whole `RunE` function is short, around 30 lines, you can read it without scrolling. I would say it's reasonably clear.

P.S. I do think the logging and local repo initialization needs more work. I'd like to work on it in another PR after this.